---
title: "Class12: RNASeq analysis"
author: "Junlin Ruan (PID: A17839687)"
format: gfm
toc: true
---

## Background
Today we will analyze some RNASeq data from Himes et al. on the effects of a common steroid (dexamethasone) on airway smooth muscle cells (ASN cells).

Are starting point is the "counts" data and "metadata" that contain the count values for each gene in their different experiments (i.e. cell lines with or without the drug).

## Data Import

```{r}
# Complete the missing code
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <- read.csv("airway_metadata.csv")
```

Let's jave a wee peak at these objects:
```{r}
head(counts)
```

> Q1. How many genes are in this dataset?

```{r}
nrow(counts)
```

> Q. How many different experiments (columns in counts or rows in metadata) are there?

```{r}
ncol(counts)
```

```{r}
nrow(metadata)
```

```{r}
metadata
```

> Q2. How many 'control' cell lines we have?

```{r}
sum(metadata$dex == "control")
```

## Toy differential gene expression

To start our analysis let's calculate the mean counts for all genes in the "control" experiments.

1. Extract all "control" columns from the `counts` object
2. Calculate the mean for all rows (i.e. genes) of these "control" columns

3-4. Do the same for "treated" 
5. Compare these `countrol.mean` and `treated.mean` values


```{r}
control.ids <- metadata$dex == "control"
control.counts <- counts[ , control.ids]
```

> Q3. How would you make the above code in either approach more robust? Is there a function that could help here?

```{r}
control.means <- rowMeans(control.counts)
```

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.ids <- metadata$dex == "treated"
treated.counts <- counts[ , treated.ids]
treated.means <- rowMeans(treated.counts)
```


Store these together for ease of bookkeeping as `meancounts`

```{r}
meancounts <- data.frame(control.means, treated.means)
head(meancounts)
```

> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples.

Make a plot of control vs treated mean values for all genes

```{r}
plot(meancounts)
```
> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

```{r}
library(ggplot2)
ggplot(meancounts, aes(x = control.means, y = treated.means)) +
  geom_point()
```
> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

Make this a log log plot

```{r}
plot(meancounts, log = "xy")
```

We often talk metrics like "log2 fold-change"

```{r}
# treated/control
log2(10/10)
```

```{r}
log2(10/20)
```

```{r}
log2(20/10)
```

```{r}
log2(40/10)
log2(10/40)
```
let's calculate the log2 fold change for our treated over control mean counts.
```{r}
meancounts$log2fc <- log2(meancounts$treated.means / meancounts$control.means)
```

```{r}
head(meancounts)
```

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The arr.ind=TRUE argument will clause which() to return both the row and column indices (i.e. positions) where there are TRUE values.
Calling unique() will ensure we donâ€™t count any row twice if it has zero entries in both samples.

A common "rule of thumb" is a log2 fold change cutoff of +2 and -2 to call genes "Up regulated" or "Down regulated".

Number of "up" genes
```{r}
sum(meancounts$log2fc >= +2, na.rm = T)
```

Number of "down" genes at -2 threshold
```{r}
sum(meancounts$log2fc <= -2, na.rm = T)
```
> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?
> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?

```{r}
up.ind <- meancounts$log2fc > 2
down.ind <- meancounts$log2fc < (-2)
sum(up.ind, na.rm = T)
sum(down.ind, na.rm = T)
```
> Q10. Do you trust these results? Why or why not?

Fold change can be large (e.g. >>two-fold up- or down-regulation) without being statistically significant (e.g. based on p-values). So we have to determine if the differences are significant first before drawing conclusions.

## DESeq2 analysis

Let's do this analysis properly and keep our inner stats nerd happy - i.e. are the differences we see between drug and no drug significant given the replicate experiments.

```{r, message=FALSE}
library(DESeq2)
```

For DESeq analysis we need three things

- count values (`contData`)
- metadata telling us about the columns in `countData` (`colData`)
- design of the experiment (i.e. what do you want to compare)


Our first function from DESeq2 will setup the input required for analysis by storing all these 3 things together.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex)
```

The main function in DESeq2 that runs the analysis is called `DESeq()`

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
head(res)
```

## Volcano Plot

This is common summary result figure from these types of experiments and plot the log2 fold-change vs the adjusted p-value.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v = c(-2,2), col = "red")
abline(h = -log(0.05), col = "blue")
```

## Save our results

```{r}
write.csv(res, file="my_results.csv")
```

## Add gene annotation

To help make sense of our results and communicate them to other folks we need to add some more annotation to our main `res` object.

We will use two bioconductor packages to first map IDs to different formats including the classic gene "symbol" gene name.

I will install these with the following commands:
`BiocManager::install("AnnotationDbi")`
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```

Let's see what is in `org.Hs.eg.db` with the `columns()` function:

```{r}
columns(org.Hs.eg.db)
```

We can translate or "map" IDs between any of these 26 databases using the `mapIds()` function.

```{r}
res$symbol <- mapIds(keys = row.names(res), # our current IDs
                     keytype = "ENSEMBL",   # the format of our IDs
                     x = org.Hs.eg.db,      # where to get the mappings from
                     column = "SYMBOL")      # the format/DB to map to
                    
head(res)
```

Add the mappings for "GENENAME" and "ENTREZID" and store as `res$genename` and `res$entrez`

```{r}
res$genename <- mapIds(keys = row.names(res), 
                     keytype = "ENSEMBL",
                     x = org.Hs.eg.db, 
                     column = "GENENAME") 
res$entrez <- mapIds(keys = row.names(res), 
                     keytype = "ENSEMBL",
                     x = org.Hs.eg.db, 
                     column = "ENTREZID") 
head(res)
```

## Pathway analysis

There are lots of bioconductor packages to do this type of analysis. For now let's just try one called **gauge** again we need ot install this if we don't have it already.

```{r, message = FALSE}
library(gage)
library(gageData)
library(pathview)
```

To use **gage** I need two things

- a named vector of fold-change values for our DEGs (our geneset of interest)
- a set of pathways or genesets to use for annotation

```{r}
x <- c("barry" = 5, "lisa" = 10)
x
```

```{r}
names(x) <- c("low", "high")
x
```

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```

```{r}
data("kegg.sets.hs")

keggres = gage(foldchanges, gsets = kegg.sets.hs)
```

In our results object we have:
```{r}
attributes(keggres)
```
```{r}
head(keggres$less, 5)
```

Let's look at one of these pathways (hsa05310 Asthma) with our genes colored up so we can see the overlap

```{r}
pathview(pathway.id = "hsa05310", gene.data = foldchanges)
```

Add this pathway figure to our lab report

![](hsa05310.pathview.png)

## Save our main results

```{r}
write.csv(res, file = "myresults_annotated.csv")
```

